app/toolbox-server.hs

```diff
@@ -0,0 +1,93 @@
+module Main (main) where
+
+import Concur.Core
+  ( Widget,
+    liftSTM,
+  )
+import Concur.Replica
+  ( div,
+    pre,
+    runDefault,
+    text,
+  )
+import Control.Applicative ((<|>))
+import Control.Concurrent (forkIO)
+import Control.Concurrent.STM
+  ( TVar,
+    atomically,
+    newTVar,
+    readTVar,
+    retry,
+    writeTVar,
+  )
+import Data.Map (Map)
+import qualified Data.Map as M
+import Data.Text (Text)
+import qualified Data.Text as T
+import Replica.VDOM.Types (HTML)
+import Toolbox.Comm
+  ( receiveObject,
+    runServer,
+  )
+import Prelude hiding (div)
+
+main :: IO ()
+main = do
+  var <- atomically $ newTVar Nothing
+  _ <- forkIO $ listener var
+  webServer var
+
+type ModuleMessages = Map Text Text
+
+listener :: TVar (Maybe (Int, ModuleMessages)) -> IO ()
+listener var =
+  runServer
+    "/tmp/ghc-build-analyzer.ipc"
+    ( \sock -> do
+        (modName, msg) <- receiveObject sock
+        -- threadDelay 500_000
+        updateModuleMessages var (modName, msg)
+    )
+
+updateModuleMessages :: TVar (Maybe (Int, ModuleMessages)) -> (Text, Text) -> IO ()
+updateModuleMessages var (modName, msg) = do
+  mmap <- atomically $ readTVar var
+  case mmap of
+    Nothing -> do
+      let m = M.singleton modName msg
+      atomically $ writeTVar var (Just (1, m))
+    Just (i, m) -> do
+      let m' = M.insert modName msg m
+      atomically $ writeTVar var (Just (i + 1, m'))
```

Comment:
I'm curious about the intent on the two `atomically` calls here. With two threads, we could conceivably have overwriting of values.

| T | Thread #1 | Thread #2 |
|-|-|-|
| 0 | Enter `updateModuleMessages` | | 
| 1 | `readTVar var` completes with `Nothing` | |

In The Good Case, we have:


| T | Thread #1 | Thread #2 |
|-|-|-|
| 2 | Write `TVar` completes | | 
| 3 |  | Read `tvar` completes with `Just` the result |
| 4 | | `writeTVar` completes successfully, results are kept |

But the Bad Case looks like:

| T | Thread #1 | Thread #2 |
|-|-|-|
| 2 |  | Enter `updateModuleMessages` | 
| 3 | | `readTVar var` completes with `Nothing` |
| 4 | `writeTVar` occurs | | 
| 5 | | `writeTVar` occurs, overwriting Thread 1 |

I think using `modifyTVar` would probably work to avoid this potential concurrency issue.

Reply:
Right!! Thanks for catching this. 
The whole `updateInbox` should be `STM ()`.

Reply:
updated.